### PR TITLE
cni: cancel sandbox ADD when the pod is deleted

### PR DESCRIPTION
**Pull request by dcbw**
Runtimes have no way to cancel an ongoing ADD request. So if a pod is added, then very quickly deleted, our CNI may be uselessly waiting up to 20 seconds for OVS flows that will never appear.

Track sandbox ADD requests in the CNI server and allow them to be canceled by a pod watch in the Node.

@trozet

### CODE REVIEW

**Comment by trozet** [comment]
@dcbw here is the current failure:

Nov 12 14:54:28 ovn-worker kubelet[421]: E1112 14:54:28.509467     421 pod_workers.go:191] Error syncing pod a868d5de-6ac8-46e6-bd87-fafec0cbeb00 ("local-path-provisioner-78776bfc44-bx8n2_local-path-storage(a868d5de-6ac8-46e6-bd87-fafec0cbeb00)"), skipping: failed to "CreatePodSandbox" for "local-path-provisioner-78776bfc44-bx8n2_local-path-storage(a868d5de-6ac8-46e6-bd87-fafec0cbeb00)" with CreatePodSandboxError: "CreatePodSandbox for pod \"local-path-provisioner-78776bfc44-bx8n2_local-path-storage(a868d5de-6ac8-46e6-bd87-fafec0cbeb00)\" failed: rpc error: code = Unknown desc = failed to setup network for sandbox \"7ef85ddb9febab8c0d84a42cd3ccd98c49fc206cef5dc87b6f1db85632ef9e6a\": CNI request failed with status 400: 'pod sandbox not found for ID 7ef85ddb9febab8c0d84a42cd3ccd98c49fc206cef5dc87b6f1db85632ef9e6a\n'"

Nov 12 14:54:38 ovn-worker kubelet[421]: E1112 14:54:38.430954     421 remote_runtime.go:113] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to setup network for sandbox "9e26335720ffa809a61a91c28311bc5b465da2236545aea92461f45059e04252": CNI request failed with status 400: 'pod sandbox not found for ID 9e26335720ffa809a61a91c28311bc5b465da2236545aea92461f45059e04252

**Comment by dcbw** [comment]
@trozet yeah, because I'm not convinced ovnkube-node is able to talk to the runtime yet... debugging.

File: go-controller/pkg/node/node.go [hunk 1]
```diff
@@ -279,12 +280,23 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 	n.WatchEndpoints()
 
 	// start the cni server
-	cniServer := cni.NewCNIServer("", kclient.KClient)
-	err = cniServer.Start(cni.HandleCNIRequest)
+	n.cniServer = cni.NewCNIServer("", kclient.KClient)
```
Comment (aojea) [inline go-controller/pkg/node/node.go:283]: is not better to plumb the informer down?
is just replacing the lister here by the informer 
https://github.com/ovn-org/ovn-kubernetes/pull/1832
and also we keep the code independent

Comment (dcbw) [reply go-controller/pkg/node/node.go:283]: @aojea Wouldn't we have to do a new setup for the factory using that informer or something?

File: go-controller/pkg/cni/ovs.go [hunk 2]
```diff
@@ -170,6 +171,12 @@ func waitForPodFlows(mac string, ifAddrs []*net.IPNet, ifaceName, ifaceID string
 				return false, nil
 			}
 		}
+
+		// check for cancelation
+		if atomic.LoadUint32(cancel) > 0 {
```
Comment (aojea) [inline go-controller/pkg/cni/ovs.go:176]: we should use a `context` and wait in parallel, this still waits for the ovs operations that may hang or take time

Comment (trozet) [reply go-controller/pkg/cni/ovs.go:176]: could we wrap the entire cmdAdd in the context and cancel any part of it instead of just this function?

File: go-controller/pkg/node/node.go [hunk 3]
```diff
@@ -279,12 +280,23 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
...
+	err = n.cniServer.Start(cni.HandleCNIRequest)
+	// Watch pods so we can cancel CNI requests when pods are deleted
+	n.WatchPods()
```
Comment (trozet) [inline go-controller/pkg/node/node.go:286]: I'm guessing we already filter the pods just for this node right?

File: go-controller/pkg/node/node.go (hunk 1 above)
Comment (aojea) [reply go-controller/pkg/node/node.go:283]: thinking out loud, is not that the CNIserver is going to be decoupled of the node, right?

File: go-controller/pkg/node/node.go (hunk 3 above)
Comment (aojea) [reply go-controller/pkg/node/node.go:286]: yeah, that informer has  a selector for the node

Comment (abhat) [reply go-controller/pkg/node/node.go:286]: I don't think so. WatchPods() is watching all pods afaict. This is the same watcher as the one we run on master afaik.

File: go-controller/pkg/cni/ovs.go (hunk 2 above)
Comment (aojea) [reply go-controller/pkg/cni/ovs.go:176]: yeah, like having a global map with the ctxs, so we just cancel from the top and it cascade the cancel down

File: go-controller/pkg/node/node.go (hunk 3 above)
Comment (aojea) [reply go-controller/pkg/node/node.go:286]: https://github.com/ovn-org/ovn-kubernetes/blob/23bda5e9fdda405ef9aa33a3cb0cfebecf8155a0/go-controller/pkg/factory/factory.go#L219-L229

Comment (aojea) [reply go-controller/pkg/node/node.go:286]: I think that casey has changed it recently,

**Comment by dcbw** [comment]
@trozet @aojea @abhat PTAL. Switched up per Tim's suggestion to use a Context object and cancel that rather than an atomic.

File: go-controller/pkg/node/node.go (hunk 1 above)
Comment (dcbw) [reply go-controller/pkg/node/node.go]: See if the current code looks like what you expect?

File: go-controller/pkg/cni/ovs.go (hunk 2 above)
Comment (dcbw) [reply go-controller/pkg/cni/ovs.go]: Done.

File: go-controller/pkg/node/node.go (hunk 3 above)
Comment (dcbw) [reply go-controller/pkg/node/node.go]: Yeah, Casey made it only watch pods on the node.

File: go-controller/pkg/cni/cniserver.go [hunk 4]
```diff
@@ -51,15 +56,45 @@ func NewCNIServer(rundir string, kclient kubernetes.Interface) *Server {
...
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			s.cancelEarliestPodAdd(pod)
+		},
+	}, nil)
+
 	return s
 }
 
+// cancelEarliestPodAdd requests that the earliest outstanding add operation for a given
```
Comment (aojea) [inline go-controller/pkg/cni/cniserver.go:77]: `Earliest` or `earlier than`? 
There should be only on ADD active, no?

```
	for _, req := range s.adds {
		if req.PodNamespace == pod.Namespace && req.PodName == pod.Name && req.timestamp.Before(earliest) {
			req.cancel()
		}
	}

File: go-controller/pkg/cni/ovs.go [hunk 5]
```diff
@@ -110,66 +111,86 @@ func ofctlExec(args ...string) (string, error) {
...
-				}
+		if !found {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func waitForPodFlows(mac string, ifAddrs []*net.IPNet, ifaceName, ifaceID string, ctx context.Context) error {
```
Comment (aojea) [inline go-controller/pkg/cni/ovs.go:181]: The pattern is to pass context as first argument

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

https://golang.org/pkg/context/

File: go-controller/pkg/cni/types.go [hunk 6]
```diff
@@ -74,15 +77,25 @@ type PodRequest struct {
...
 type Server struct {
 	http.Server
 	requestFunc cniRequestFunc
 	rundir      string
-	kclient     kubernetes.Interface
+	podLister   corev1listers.PodLister
+
+	// adds is a map of sandbox ID to PodRequest for any CNIAdd operation
+	addsLock sync.Mutex
+	adds     map[string]*PodRequest
```
Comment (aojea) [inline go-controller/pkg/cni/types.go:100]: nit. naming is hard but this is very confusing

File: go-controller/pkg/cni/ovs.go [hunk 7]
```diff
@@ -110,66 +111,86 @@ func ofctlExec(args ...string) (string, error) {
...
+	}
+	if out != ifaceID {
+		return false, fmt.Errorf("OVS sandbox port %s is no longer active (probably due to a subsequent "+
+			"CNI ADD)", ifaceName)
+	}
+
+	return true, nil
+}
+
+func doPodFlowsExist(mac string, ifAddrs []*net.IPNet) (bool, error) {
```
Comment (aojea) [inline go-controller/pkg/cni/ovs.go:129]: Food for thought, is this a hard requirement or can we relax it and hand over the responsibility to OVN/OVN controller to reconcile on this and make that the flows are set up correctly... flows must be there sooner or later, or can it happen that it never make it?

File: go-controller/pkg/cni/cniserver.go [hunk 8]
```diff
@@ -155,6 +212,13 @@ func (s *Server) handleCNIRequest(w http.ResponseWriter, r *http.Request) {
 			klog.Warningf("Error writing %s HTTP response: %v", req.Command, err)
 		}
 	}
+
+	if req.Command == CNIAdd {
```
Comment (aojea) [inline go-controller/pkg/cni/cniserver.go:216]: should the CNIDel  clear the cache too?

File: go-controller/pkg/cni/ovs.go [hunk 9]
```diff
@@ -110,66 +111,86 @@ func ofctlExec(args ...string) (string, error) {
...
+	for _, query := range queries {
+		found := false
+		// Look for a match in any table to be considered success
+		for _, table := range query.tables {
+			queryStr := fmt.Sprintf("table=%d,%s", table, query.match)
+			// ovs-ofctl dumps error on stderr, so stdout will only dump flow data if matches the query.
+			stdout, err := ofctlExec("dump-flows", "br-int", queryStr)
+			if err == nil && len(stdout) > 0 {
+				found = true
+				break
```
Comment (aojea) [inline go-controller/pkg/cni/ovs.go:170]: based on the comment 
> 		// Look for a match in any table to be considered success

I think this should return directly, [because the break in golang only breaks the inner loop](https://play.golang.org/p/lvP9tqdWNVy), so it will set found to false and check again, if that is the purpose the comment should be to match in ALL tables

otherwise we should rewrite it to something like

```
for _, query := range queries {
		// Look for a match in any table to be considered success
		for _, table := range query.tables {
			queryStr := fmt.Sprintf("table=%d,%s", table, query.match)
			// ovs-ofctl dumps error on stderr, so stdout will only dump flow data if matches the query.
			stdout, err := ofctlExec("dump-flows", "br-int", queryStr)
			if err == nil && len(stdout) > 0 {
                                  return true, nil
                        }
            }
}
// not flows found
return false, nil
}
```

File: go-controller/pkg/cni/cni.go [hunk 10]
```diff
@@ -199,3 +179,28 @@ func (pr *PodRequest) getCNIResult(podInterfaceInfo *PodInterfaceInfo) (*current
...
+			}
+			annotations = pod.ObjectMeta.Annotations
+			if _, ok := annotations[util.OvnPodAnnotationName]; ok {
+				return true, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get pod %s/%s annotations: %v", name, namespace, err)
+	}
```
Comment (aojea) [inline go-controller/pkg/cni/cni.go:204]: I don't know if we should use the `wait.PollImmediate()` here and do ourselves the loop, seems clearer to me:
https://play.golang.org/p/tBJh8zn2NKQ i.e. (is an example, I have problems with indentation in comments)
```go
func getAnnotationsOnPod(ctx context.Context, podLister corev1listers.PodLister, namespace, name string) (map[string]string, error) {

var annotations map[string]string
timeout := time.After(30 * time.Second)
pollInt := 200*time.Millisecond
for {
   select {
         case <-ctx.Done():
		return nil, fmt.Errorf("pod operation canceled")
         case <- timeout:
                  return nil, fmt.Errorf("failed to get pod %s/%s annotations: %v", name, namespace, err)
          default:
                        pod, err := podLister.Pods(namespace).Get(name)
			if err != nil {
				return nil, err
			}
			annotations = pod.ObjectMeta.Annotations
			if _, ok := annotations[util.OvnPodAnnotationName]; ok {
				return annotations, nil
			}
            }
            time.Sleep(pollInt)
       }
}

File: go-controller/pkg/cni/ovs.go [hunk 11]
```diff
@@ -110,66 +111,86 @@ func ofctlExec(args ...string) (string, error) {
...
+	return true, nil
+}
+
+func waitForPodFlows(mac string, ifAddrs []*net.IPNet, ifaceName, ifaceID string, ctx context.Context) error {
+	return wait.PollImmediate(200*time.Millisecond, 20*time.Second, func() (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, fmt.Errorf("pod operation canceled")
+		default:
+			if ok, err := isIfaceIDSet(ifaceName, ifaceID); !ok {
```
Comment (aojea) [inline go-controller/pkg/cni/ovs.go:187]: We don't need to use the  `wait.PollImmediate()`, see https://github.com/ovn-org/ovn-kubernetes/pull/1830/files#r522848053

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (aojea) [reply go-controller/pkg/cni/cniserver.go:216]: or cancel the pending Adds, if they exist?

File: go-controller/pkg/cni/ovs.go (hunk 9 above)
Comment (aojea) [reply go-controller/pkg/cni/ovs.go:170]: .. or maybe my english is wrong and it means precisely to match in ALL tables :thinking:

File: go-controller/pkg/cni/cni.go (hunk 10 above)
Comment (trozet) [reply go-controller/pkg/cni/cni.go:204]: your example is what I had in mind as well

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (trozet) [reply go-controller/pkg/cni/cniserver.go:216]: we cant get a CNIDel for a CNIAdd that is already happening, I believe those events are serial

File: go-controller/pkg/cni/ovs.go (hunk 7 above)
Comment (trozet) [reply go-controller/pkg/cni/ovs.go:129]: we cannot relax the check, and did you read the bz below? :)

File: go-controller/pkg/cni/ovs.go (hunk 9 above)
Comment (trozet) [reply go-controller/pkg/cni/ovs.go:170]: the break on the inner loop is correct. We want to ensure all queries are successful. If we find any match of the query on one of the tables in the inner loop, that query is considered successful.

File: go-controller/pkg/cni/cniserver.go (hunk 4 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:77]: There may be >= 0 sandboxes for the same Namespace+Name. When a sandbox fails, or the pod is deleted, the sandbox is put on the GC list and deleted shortly after (up to 3 minutes IIRC). During that time, another sandbox may be started for the same pod, or a new instance of the pod.

Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:77]: Since we don't have the Pod UID here, we just assume that the delete should cancel the earliest sandbox, because that sandbox is superceded by the newer one anyway.

File: go-controller/pkg/cni/types.go (hunk 6 above)
Comment (trozet) [reply go-controller/pkg/cni/types.go:100]: maybe pendingCNIAdds ?

File: go-controller/pkg/cni/ovs.go (hunk 7 above)
Comment (dcbw) [reply go-controller/pkg/cni/ovs.go:129]: Dumitru is working on a better method right now actually; but that won't be available in acutal builds for a while. We've been using this workaround for a while.

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:216]: A CNI ADD and DEL for the *same sandbox* are always serialized. We do not care about DEL for two reasons: (1) Delete doesn't touch anything that other sandboxes would care about, and (2) DEL should clean up as much as it can. So we can (and should) just let delete run since it won't affect anything else, and we should let delete run because it still needs to clean up stuff that a completed ADD may have allocated. If we canceled the DEL, it might stop before cleaning everything up.

Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:216]: Also, we only want to clean up this specific sandbox request, not one for another sandbox of the same namespace/name since htat one might be the active one.

Comment (abhat) [reply go-controller/pkg/cni/cniserver.go:216]: Would it make sense to call cancelPodRequest for the previous ADD when the cniAdd happens for the same `namespace/name` again? If so, can we then get rid of the timestamp logic altogether since there will only be one pending add at any point.

Comment (abhat) [reply go-controller/pkg/cni/cniserver.go:216]: It might entail caching the sandbox-id of the active request in a map[namespace\name]->sandbox_id

File: go-controller/pkg/cni/cniserver.go (hunk 4 above)
Comment (aojea) [reply go-controller/pkg/cni/cniserver.go:77]: can you add a comment with that explanation it will be very useful

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (aojea) [reply go-controller/pkg/cni/cniserver.go:216]: can we add a comment explaining this too?
https://github.com/ovn-org/ovn-kubernetes/pull/1830/files#r522996143

File: go-controller/pkg/cni/ovs.go (hunk 7 above)
Comment (aojea) [reply go-controller/pkg/cni/ovs.go:129]: sorry, let me rephrase, what I'm trying to mean is that I see this as a 2 part process, 
1. binding the port in the node, so the local plumbing is ready
2. certify thath OVN has everything set, OVN install the flows and everything it need to provide connectivity to thepod
The bz talks about a way to check 2 from the Node, what I'm asking is, if we can skip 2. once 1. is ready, because it may possible to ask the OVN-Kube controller to reconcile 2 , checking in a loop until OVN finish (naive question, don't know the internals of the OVN flows setup)

File: go-controller/pkg/cni/ovs.go (hunk 9 above)
Comment (aojea) [reply go-controller/pkg/cni/ovs.go:170]: so, it was my english , sorry :sweat_smile: 
> // Look for a match in any table to be considered success

File: go-controller/pkg/cni/types.go (hunk 6 above)
Comment (aojea) [reply go-controller/pkg/cni/types.go:100]: or cmdAddMap, just that `adds`  and `addsLock` seems is trying to `add` something to

File: go-controller/pkg/cni/cniserver.go (hunk 4 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:77]: Done.

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go:216]: > Would it make sense to call cancelPodRequest for the previous ADD when the cniAdd happens for the same `namespace/name` again? If so, can we then get rid of the timestamp logic altogether since there will only be one pending add at any point.

@abhat yes, this makes sense, but we can't do this without keeping sandboxID as the map key I think. Otherwise removing from the map becomes more complicated if the request got canceled vs. finished normally.

Comment (abhat) [reply go-controller/pkg/cni/cniserver.go:216]: Well, I didn't mean to get rid of the logic when the informer cache del happens when DeleteFunc is called. I am saying we can simplify the deletion logic by not relying on timestamps and instead assume that there will be only one outstanding request at any point. If the second add happens, we can cancel the pending add. So in some ways, there are two possibilities for canceling the pod request - a) pod del update comes via the informer cache update b) a cniAdd happens for the same pod (`namespace/name`). 

When the request finishes normally, you have all the necessary data to remove from the map at hand right? When the request gets cancelled via the informer del, you still only need to know the `namespace\name` for the pod if that is the key into the map. Isn't that right?

File: go-controller/pkg/cni/ovs.go (hunk 11 above)
Comment (dcbw) [reply go-controller/pkg/cni/ovs.go]: Fixed.

File: go-controller/pkg/cni/cni.go (hunk 10 above)
Comment (dcbw) [reply go-controller/pkg/cni/cni.go]: Fixed.

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go]: @abhat not quite. The problem I see is that the existing sandbox request would need to be cleared from two places: (1) at the end of handleCNIRequest() when the request has finished normally, and (2) a sub-function of handleCNIRequest() when a newer request comes in. These two things are in two different goroutines though.

One way I can think of is to check if the map value is the same request object as is being removed:

```
func makeRequestKey(namespace, name string) string {
	return req.PodNamespace + "/" + req.PodName)
}

func (s *Server) startSandboxRequest(req *PodRequest) error {
	// Only sandbox add requests are tracked because only adds need
	// to be canceled when a newer request comes in.
	if req.Command == CNIAdd {
		s.runningSandboxAddsLock.Lock()
		defer s.runningSandboxAddsLock.Unlock()
		key := makeRequestKey(req.PodNamespace, req.PodName)
		if _, ok := s.runningSandboxAdds[key]; ok {
			req.cancel()
		}
		s.runningSandboxAdds[key] = req
	}
	return nil
}

func (s *Server) finishSandboxRequest(req *PodRequest) {
	if req.Command == CNIAdd {
		s.runningSandboxAddsLock.Lock()
		defer s.runningSandboxAddsLock.Unlock()
		key := makeRequestKey(req.PodNamespace, req.PodName)
		if existing, ok := s.runningSandboxAdds[key]; ok && existing == req {
			delete(s.runningSandboxAdds, key)
		}
	}
}
```

but that seems pretty ugly and is not a pattern I've commonly seen.

Comment (abhat) [reply go-controller/pkg/cni/cniserver.go]: You can simplify it further - pass the `cancel` vs `delete` request semantics as an argument in the two different scenarios.

Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go]: And for the record, this doesn't work if we get a ADD1, ADD2, Pod1 Delete as the ADD2 would have canceled the ADD1, and then the Pod1 Delete would cancel the ADD2 which is still valid.

Comment (abhat) [reply go-controller/pkg/cni/cniserver.go]: if we can get the sandbox id when Pod1 Delete happens and if it doesn't match the ADD2's sandbox-id, we can skip?

File: go-controller/pkg/cni/cniserver.go [hunk 12]
```diff
@@ -137,10 +138,36 @@ func cniRequestToPodRequest(cr *Request) (*PodRequest, error) {
...
 	}
+
 	req.CNIConf = conf
+	req.timestamp = time.Now()
 	return req, nil
 }
 
+func (s *Server) startSandboxRequest(req *PodRequest) error {
+	// Only sandbox add requests are tracked because only adds need
+	// to be canceled when a newer request comes in.
```
Comment (trozet) [inline go-controller/pkg/cni/cniserver.go:189]: or when a delete comes in. CNI Del doesn't need to be cancelled and may proceed as usual.

File: go-controller/pkg/cni/cni.go [hunk 13]
```diff
@@ -180,23 +180,26 @@ func (pr *PodRequest) getCNIResult(podInterfaceInfo *PodInterfaceInfo) (*current
...
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("canceled waiting for annotations")
+		case <-timeout:
+			return nil, fmt.Errorf("timed out waiting for annotations")
+		default:
+			pod, err := podLister.Pods(namespace).Get(name)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get annotations: %v", err)
```
Comment (trozet) [inline go-controller/pkg/cni/cni.go:195]: maybe a better error message here like "unable to find pod (probably deleted)" or something

**Review by trozet — APPROVED** [review]
lgtm other than some minor nits on the comments

File: go-controller/pkg/cni/cni.go (hunk 13 above)
Comment (dcbw) [reply go-controller/pkg/cni/cni.go:195]: @trozet since we're asking the informer cache now, rather than the API, we'll get a descriptive error from:

```
func (s podNamespaceLister) Get(name string) (*v1.Pod, error) {
	obj, exists, err := s.indexer.GetByKey(s.namespace + "/" + name)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, errors.NewNotFound(v1.Resource("pod"), name)
	}
	return obj.(*v1.Pod), nil
}
```

File: go-controller/pkg/cni/cniserver.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/cni/cniserver.go]: @abhat Pod1 Delete would be an api event, not a CNI event, so we don't get the sandbox ID with it.

**Comment by abhat** [comment]
LGTM
